### ISSUE TITLE
R-Julia communication, discussion

### ISSUE THREAD

**Issue by s3alfisc** [s3alfisc/wildboottestjlr#14]
Hi David, 

This is a rather long entry on how to manage the communication between R and Julia via JuliaCall that aims to illustrate some advantages and pitfalls. 

It is my understanding that there are two ways to call a Julia function from R via JuliaCall and pass R objects to it. 

Let's assume our Julia function of interest is 
```
julia_func(x = 0, y = 0, z = 0) = x + y - z
```
which we can make callable from R by running (and is of course just a stand-in for `wildboottest.jl`, where `z` might e.g. stand for all function arguments that are only relevant for IV). 

```
library(JuliaCall) 
julia_setup("C:/Users/alexa/AppData/Local/Programs/Julia-1.6.3/bin")
julia_command("julia_func(x = 0, y = 0, z = 0) = x + y - z")
julia_exists("julia_func")
```
Now we can evaluate `julia_func()` from R at given values for `x = 1, y = 1, z = 2` by either running 

```
julia_call("julia_func", 1, 1, 2)
```
This is nice and clean - and in particular, the Julia environment is not polluted with values for the function arguments `x, y, z`. Unfortunately, `julia_call()` does not work with abstract types. Further, all function arguments need to be listed in the correct order.

An alternative is to send `x = 1, y = 1, z = 2` directly to Julia via `julia_assign()`:

```
x <- y <- 1
z <- 2
JuliaCall::julia_assign("x",x)
JuliaCall::julia_assign("y",y)
JuliaCall::julia_assign("z",z)
JuliaCall::julia_eval("julia_func(x, y, z)").
```
Here, if a variable would be a `special type` defined in `WildBootTests.jl`, one could communicate that to the Julia session via `julia_eval`
```
#don't run 
#julia_eval("WildBootTests.x")
```

This is what currently happens in `wildboottestjlr`. Unfortunately, this method is quite dangerous, as it pollutes the Julia environment. Here's an example: 

Suppose that, after using `julia_assign()` to send R objects to Julia and having evaluated `julia_func()` as illustrated above, I am interested in evaluating julia_func() at some of its starting values: `x = y = 0; z = 2`.
(The equivalent in the wildboottestjlr world would be an OLS bootstrap with fixed effect, and then one without one - the fixed effect variable would still be defined).

Via `julia_assign()`, I would run
```
z <- 2
JuliaCall::julia_assign("z",z)
JuliaCall::julia_eval("julia_func(x, y, z)")
# 0
```
which evaluates to zero (because x and y are still floating around in the global namespace), but should be -2.

The obvious work-around is to 're-assign' all these values 

```
x <- y <- 0
z <- 2
JuliaCall::julia_assign("x",x)
JuliaCall::julia_assign("y",y)
JuliaCall::julia_assign("z",z)
JuliaCall::julia_eval("julia_func(x, y, z)")
```

In short: both approaches require me to specify all function arguments of `wildboottest.jl`, even if I want to use the function's defaults. Option 1 is cleaner, as no objects remain in the Julia environment, but does not handle special Julia types.
